src/adaptivity.jl

```diff
@@ -115,149 +107,153 @@ function redistribute(mesh, Nsub_star::Int, ŝ::AbstractVector{T}, dt) where {T
             k += 1
         end
     end
-    mesh_new[end] = mesh[end]
-    return mesh_new
+    cache.mesh[end] = mesh[end]
+    append!(cache.mesh_dt, Nsub_star - N)
+    diff!(cache.mesh_dt, cache.mesh)
+    return cache
 end
 
 """
-    half_mesh(mesh)
-
-The input mesh has length of n+1
+    half_mesh!(mesh, mesh_dt)
+    half_mesh!(cache::MIRKCache)
 
-Divide the original subinterval into two equal length subinterval.
+The input mesh has length of `n + 1`. Divide the original subinterval into two equal length
+subinterval. The `mesh` and `mesh_dt` are modified in place.
 """
-@views function half_mesh(mesh::AbstractVector{T}) where {T}
+function half_mesh!(mesh::Vector{T}, mesh_dt::Vector{T}) where {T}
```

Comment:
only Vector?

Reply:
Mesh is controlled internally by us. Irrespective of what the input types are we always construct a Vector. We can relax it, but I am not sure how many types declare stuff like `resize!`

Reply:
Won't the mesh need to be a GPU vector for the GPU stuff to work?

Reply:
Not currently. But again, the code right now is very slow on GPU.

The main issue is we iterate over the mesh to redistribute the new points, so we need to assume fast scalar indexing. We will have to figure out a way to batch the entire computation (which technically is trivial to write a kernel for).